Common/include/parallelization/mpi_structure.hpp

```diff
@@ -0,0 +1,687 @@
+
+    MinRankError = Size;
+    MPI_Win_create(&MinRankError, sizeof(int), sizeof(int), MPI_INFO_NULL, currentComm, &winMinRankError);
+    winMinRankErrorInUse = true;
+  }
+
+  static inline void Buffer_attach(void* buffer, int size) { MPI_Buffer_attach(buffer, size); }
+
+  static inline void Buffer_detach(void* buffer, int* size) { MPI_Buffer_detach(buffer, size); }
+
+  static inline void Comm_rank(Comm comm, int* rank) { MPI_Comm_rank(comm, rank); }
+
+  static inline void Comm_size(Comm comm, int* size) { MPI_Comm_size(comm, size); }
+
+  static inline void Finalize() {
+    if (winMinRankErrorInUse) MPI_Win_free(&winMinRankError);
+    MPI_Finalize();
+  }
+
+  static inline void Barrier(Comm comm) { MPI_Barrier(comm); }
+
+  static inline void Abort(Comm comm, int error) { MPI_Abort(comm, error); }
+
+  static inline void Get_count(Status* status, Datatype datatype, int* count) {
+    MPI_Get_count(status, datatype, count);
+  }
+
+  static inline void Isend(const void* buf, int count, Datatype datatype, int dest, int tag, Comm comm,
+                           Request* request) {
+    MPI_Isend(buf, count, datatype, dest, tag, comm, request);
+  }
+
+  static inline void Irecv(void* buf, int count, Datatype datatype, int dest, int tag, Comm comm, Request* request) {
+    MPI_Irecv(buf, count, datatype, dest, tag, comm, request);
+  }
+
+  static inline void Wait(Request* request, Status* status) { MPI_Wait(request, status); }
+
+  static inline void Testall(int count, Request* array_of_requests, int* flag, Status* array_of_statuses) {
+    MPI_Testall(count, array_of_requests, flag, array_of_statuses);
+  }
+
+  static inline void Waitall(int nrequests, Request* request, Status* status) {
+    MPI_Waitall(nrequests, request, status);
+  }
+
+  static inline void Probe(int source, int tag, Comm comm, Status* status) { MPI_Probe(source, tag, comm, status); }
+
+  static inline void Send(const void* buf, int count, Datatype datatype, int dest, int tag, Comm comm) {
+    MPI_Send(buf, count, datatype, dest, tag, comm);
+  }
+
+  static inline void Recv(void* buf, int count, Datatype datatype, int dest, int tag, Comm comm, Status* status) {
+    MPI_Recv(buf, count, datatype, dest, tag, comm, status);
+  }
+
+  static inline void Bcast(void* buf, int count, Datatype datatype, int root, Comm comm) {
+    MPI_Bcast(buf, count, datatype, root, comm);
+  }
+
+  static inline void Bsend(const void* buf, int count, Datatype datatype, int dest, int tag, Comm comm) {
+    MPI_Bsend(buf, count, datatype, dest, tag, comm);
+  }
+
+  static inline void Reduce(const void* sendbuf, void* recvbuf, int count, Datatype datatype, Op op, int root,
+                            Comm comm) {
+    MPI_Reduce(sendbuf, recvbuf, count, datatype, op, root, comm);
+  }
+
+  static inline void Allreduce(const void* sendbuf, void* recvbuf, int count, Datatype datatype, Op op, Comm comm) {
+    MPI_Allreduce(sendbuf, recvbuf, count, datatype, op, comm);
+  }
```

Comment:
here now const sendbuf. Now as long as nobody writes code before the MPI_Allreduce call that changes sendbuf we were good in terms of correctness before, right? const probably also adds a little performance gain here? 

Are those the the only changes to this hpp+inl merge?

Reply:
Before if we had something like:
```c++
struct A {
  int a;
  int fun() const {
    SU2_MPI::Allreduce(&a,...);
  }
};
```
It would not work because "a" is const and the compiler will not let you take a non-const address.
As we make more parts of the code const correct and use containers instead of bare pointers, these non const "sendbuffs" would be a problem.

Reply:
Ah! and yes those are the only things that changed.